### PR TITLE
Add breaking change about host namespace causing mapping issues

### PR DESCRIPTION
**Pull request by dedemorton**
Fixes https://github.com/elastic/beats/issues/7368

I found the original document hard to follow (had to read it through multiple times), so I tried to rewrite it my own words based on what I understood after reading the document. We need to connect more of the dots for users who aren't familiar with this problem.

We don't have enough section levels in asciidoc to properly format this, so I've fudged some sections (don't tell the other writers).

Not sure it's 100% there, but here you go.

UPDATE: Please make sure you respond to the questions I've raised for reviewers.

### CODE REVIEW

File: libbeat/docs/breaking.asciidoc [hunk 1]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+is defined in the Elasticsearch index template even if the processor is not
+used.
+
+We've also added a `host.name` field to all events sent by Beats. This field
+prevents the Beats input plugin in Logstash from adding a default `host` field.
+(By default, the plugin adds a `host` field if the event doesn't already have
+one. We don't want the plugin to add this field because it causes a mapping
+conflict with the `host` object defined in the index template.)
+
+*What does this mean to you?*
```
Comment (dedemorton) [inline libbeat/docs/breaking.asciidoc]: Hmmm...the formatting here looks OK in the built docs, but not so good in GitHub, so I might change it.

File: libbeat/docs/breaking.asciidoc [hunk 2]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+// REVIEWERS: I describe these as mapping conflicts throughout, but the acutal
+// error message says mapper parsing exception. I'm wondering if I should use
+// that instead?
+
+This breaking change applies only to users who send Beats events to Logstash.
+
+Starting with version 6.3, Beats provides an `add_host_metadata` processor for
+adding fields, such as `host.name` and `host.id`, to Beats events. These fields
+are grouped under a `host` prefix and conform to the
+https://github.com/elastic/ecs[Elastic Common Schema (ECS)]. The `host` object
```
Comment (dedemorton) [inline libbeat/docs/breaking.asciidoc]: I hope it's OK to call this the `host` object because saying `host` field doesn't convey the idea of the type being different.

**Comment by dedemorton** [comment]
@karenzone If you have the time, can you take a look at this and offer suggestions for places where the language is confusing? Once you understand the problem, the details are easy to follow...but otherwise is a confusing situation.

File: libbeat/docs/breaking.asciidoc [hunk 3]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+=== Breaking changes in 6.3
+
+This section discusses the main changes that you should be aware of if you
+upgrade the Beats to version 6.3. Please also review the relevant
+Breaking Changes sections of the <<release-notes,release notes>>.
+
+[[breaking-changes-mapping-conflict]]
+==== New `host` namespace may cause mapping conflicts for Logstash
+
+// REVIEWERS: I describe these as mapping conflicts throughout, but the acutal
```
Comment (ruflin) [inline libbeat/docs/breaking.asciidoc]: I like mapping conflict as it is a mapping conflict :-)

File: libbeat/docs/breaking.asciidoc [hunk 4]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+`host.name` field in 6.3 indices. There are no mapping conflicts. Cross index
+searches in Elasticsearch still work as expected. 
+
+*What do you need to change?*
+
+If you searched for the `host` field previously, you now need to search for the
+`host` and `host.name` fields. Or you can adjust your queries to use
+`beat.hostname`, instead of `host`, to work across all indices. 
+
+// REVIEWERS: Would it better just to tell users to uses to change their
```
Comment (ruflin) [inline libbeat/docs/breaking.asciidoc]: `beat.hostname` is going to disappear in 7.0. So I would prefer to stear people in the direction of `host.name` if possible so they have to change it only once.

File: libbeat/docs/breaking.asciidoc [hunk 5]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+// searches to use `beat.hostname` instead of `host`?
+
+If you have visualizations in Kibana that reference the `host` field, export the
+objects by using the Kibana API or UI, change any aggregation references to use
+`beat.hostname` instead of `host`, and then re-import the objects into Kibana.
+
+// REVIEWERS: Are the above details correct? Got them from a comment Pius made
```
Comment (ruflin) [inline libbeat/docs/breaking.asciidoc]: The above should work but if someone has only 1 search based on the `host` field I would rather recommend to directly edit it in Kibana instead of exporting / importing it. The above might be better if there are lots of queries dependening on it but I think it requires more knowledge as the files have to be edited manually. 

I would propose both options.

@ppf2 WDYT?

File: libbeat/docs/breaking.asciidoc [hunk 6]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
++
+--
+IMPORTANT: If you drop the `host` fields, you cannot use the `add_host_metadata`
+processor.
+
+--
++
+With this configuration, Beats drops the `host` fields before sending the
+event to Logstash, and Logstash adds a default `host` field, as it did with
+previous Beats versions.
```
Comment (ruflin) [inline libbeat/docs/breaking.asciidoc]: Long term I think the `host` field will disappear for the beats-input-plugin in Logstash. I wonder if we should make a recommendation here to start thinking about how to migrate long term to `host.name`.

File: libbeat/docs/breaking.asciidoc [hunk 7]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+You cannot resolve the problem by dropping the `host.*` fields, as described
+earlier, because Logstash will add a default `host` field, resulting in a
+mapping conflict because the index template defines the `host` field as an
+object. 
+
+*What do you need to change?*
+
+To resolve the mapping issue, do *one* of the following:
+
+// TOOD: Should add an example of the LS filter here, don't you think?
```
Comment (ruflin) [inline libbeat/docs/breaking.asciidoc]: ++

@ph Could you provide one? 

There is a bit a difference if `host.*` is dropped by Beats or LS. If it is dropped by Beats, LS will add `host` field. If it's dropped by LS, there will be no host field at all in the final event. I wonder if we should mention that difference.

Comment (ph) [reply libbeat/docs/breaking.asciidoc]: @ruflin @dedemorton  Yes, the following example will bring to pre 6.3.0. 

```
mutate {
  remove_field => [ "[host]" ]
}
mutate {
  add_field => {
    "host" => "%{[beat][hostname]}"
   }
}
```

File: libbeat/docs/breaking.asciidoc (hunk 3 above)
Comment (ph) [reply libbeat/docs/breaking.asciidoc]: It's a mapping conflict.

File: libbeat/docs/breaking.asciidoc [hunk 8]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+Breaking Changes sections of the <<release-notes,release notes>>.
+
+[[breaking-changes-mapping-conflict]]
+==== New `host` namespace may cause mapping conflicts for Logstash
+
+// REVIEWERS: I describe these as mapping conflicts throughout, but the acutal
+// error message says mapper parsing exception. I'm wondering if I should use
+// that instead?
+
+This breaking change applies only to users who send Beats events to Logstash.
```
Comment (ph) [inline libbeat/docs/breaking.asciidoc]: And users using their own template.

File: libbeat/docs/breaking.asciidoc (hunk 7 above)
Comment (ph) [reply libbeat/docs/breaking.asciidoc]: @ruflin but taking the content of `[beat][hostname]` for host will result in the same behavior as the input had. WDYT?

File: libbeat/docs/breaking.asciidoc [hunk 9]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+https://github.com/elastic/ecs[Elastic Common Schema (ECS)]. The `host` object
+is defined in the Elasticsearch index template even if the processor is not
+used.
+
+We've also added a `host.name` field to all events sent by Beats. This field
+prevents the Beats input plugin in Logstash from adding a default `host` field.
+(By default, the plugin adds a `host` field if the event doesn't already have
+one. We don't want the plugin to add this field because it causes a mapping
+conflict with the `host` object defined in the index template.)
+
```
Comment (karenzone) [inline libbeat/docs/breaking.asciidoc]: The first sentence in the parentheses doesn't say anything new, so consider deleting it, and adding `host` to remaining sentence.

+(We don't want the plugin to add the `host` field because it causes a mapping
+conflict with the `host` object defined in the index template.)

File: libbeat/docs/breaking.asciidoc (hunk 1 above)
Comment (karenzone) [reply libbeat/docs/breaking.asciidoc]: I see what you mean about formatting being weird in GitHub. That threw me off at first. Interesting that it builds correctly in the book.

File: libbeat/docs/breaking.asciidoc [hunk 10]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+Beats, you might need to modify your searches and visualizations. See:
+
+* <<beats-template-versioned-indices>>
+
+If you *are not* using the recommended index template and versioned indices,
+see the section for your use case:
+
+* <<custom-template-non-versioned-indices>>
+* <<beats-template-non-versioned-indices>>
+
```
Comment (karenzone) [inline libbeat/docs/breaking.asciidoc]: +*What does this mean to you?*
See the info for your particular use case:
* <<link use case 1>>
* <<link use case 2>>
* <<link use case 3>>

You did a good job capturing the use case scenario in each title.  The use cases are easier to read and parse without the surrounding text.

File: libbeat/docs/breaking.asciidoc (hunk 7 above)
Comment (ph) [reply libbeat/docs/breaking.asciidoc]: I will do a check, but the _add_field_ should replace the original value.

File: libbeat/docs/breaking.asciidoc [hunk 11]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+[[beats-template-non-versioned-indices]]
+===== Use case: You use the Beats index template and your indices are not versioned
+
+In this use case, you load the Beats index template manually into Elasticsearch,
+and send your data through Logstash, but you don’t use the versioned index
+pattern to create versioned indices.
+
+You cannot resolve the problem by dropping the `host.*` fields, as described
+earlier, because Logstash will add a default `host` field, resulting in a
+mapping conflict because the index template defines the `host` field as an
```
Comment (karenzone) [inline libbeat/docs/breaking.asciidoc]: "earlier" might be problematic.  Linking to different use cases doesn't promote linear reading, but "earlier" seems to imply that you've read everything.  Here, I think we can delete the reference to earlier and just say, "You cannot resolve the problem by dropping the `host.*` fields,  because Logstash will add a default `host` field..."
Especially since what we're referencing isn't going to work in this situation anyway. :-)

File: libbeat/docs/breaking.asciidoc [hunk 12]
```diff
@@ -9,9 +9,152 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+To resolve the mapping issue, do *one* of the following:
+
+// TOOD: Should add an example of the LS filter here, don't you think?
+
+* Drop the `host.*` fields in a Logstash filter. There will be no
+`host` field at ingestion time, and the mapping conflict is avoided.
+* Or:
+** Modify the Beats index template by removing the `host.*` fields, and
+** Configure Beats to drop all `host.*` fields, as described earlier.
++
```
Comment (karenzone) [inline libbeat/docs/breaking.asciidoc]: WRT "earlier".  Same concern applies (this document isn't designed to be read linearly). 

If we're referencing this: "In the Beats config file, configure Beats to drop all `host.*` fields: ", just say that instead of referencing it.

**Comment by brandonmensing** [comment]
LGTM - nothing to add on top of the comments and todos already going on there. Thanks!

File: libbeat/docs/breaking.asciidoc (hunk 8 above)
Comment (dedemorton) [reply libbeat/docs/breaking.asciidoc]: @ph You're saying that this issue applies to Beats users who use a custom template and aren't using Logstash, correct? If they have `host` field defined in their custom template, they'll run into mapping errors, no?

File: libbeat/docs/breaking.asciidoc (hunk 9 above)
Comment (dedemorton) [reply libbeat/docs/breaking.asciidoc]: I had it that way originally, but thought being explicit might be useful for users who don't understand that LS adds the field if there's not one. This behavior isn't consistent across input plugins, so it might be worth calling out. But I agree...it's a bit redundant--that's why I made it parenthetical even though I know people hate parens. If no one disagrees, I will remove the statement.

File: libbeat/docs/breaking.asciidoc (hunk 1 above)
Comment (dedemorton) [reply libbeat/docs/breaking.asciidoc]: Well the docs use our CSS to control the size of things. GitHub shouldn't make bolded text into a big section header, but it does.

File: libbeat/docs/breaking.asciidoc (hunk 4 above)
Comment (dedemorton) [reply libbeat/docs/breaking.asciidoc]: OK, then I'll remove mention of this field.

File: libbeat/docs/breaking.asciidoc (hunk 8 above)
Comment (ph) [reply libbeat/docs/breaking.asciidoc]: @dedemorton 
It applies to the following use cases:

- FB -> LS,  using the default logstash-output-elasticsearch template.
- FB -> LS,  using their own template templates with the host defined.

File: libbeat/docs/breaking.asciidoc (hunk 12 above)
Comment (dedemorton) [reply libbeat/docs/breaking.asciidoc]: I'll add a link to the section because I want users to know that the topic describes how to do this.

File: libbeat/docs/breaking.asciidoc (hunk 7 above)
Comment (dedemorton) [reply libbeat/docs/breaking.asciidoc]: @ph See my comment ^^

**Comment by dedemorton** [comment]
I've pushed some changes based on review feedback. There are a couple of sections that still need work pending a response from reviewers:
@ppf2 - https://github.com/elastic/beats/pull/7398/files#r197348926

@ph @ruflin https://github.com/elastic/beats/pull/7398#discussion_r197524285

@tsg FYI...

File: libbeat/docs/breaking.asciidoc (hunk 7 above)
Comment (ph) [reply libbeat/docs/breaking.asciidoc]: @dedemorton this is perfectly valid, thanks for  that :)
```
mutate {
  add_field => {
    "host" => "%{[host][name]}"
   }
}```

File: libbeat/docs/breaking.asciidoc (hunk 8 above)
Comment (ph) [reply libbeat/docs/breaking.asciidoc]: @dedemorton Yes this is true, so let's keep it more specific we can adjust if people are posting on discuss.

File: libbeat/docs/breaking.asciidoc [hunk 13]
```diff
@@ -9,9 +9,155 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+If you searched for the `host` field previously, you now need to search for the
+`host` and `host.name` fields.  
+
+If you have multiple visualizations in Kibana that reference the `host` field,
+you can export the objects by using the Kibana API or UI, change any aggregation
+references to use `host.name` instead of `host`, and then re-import the
+objects into Kibana.
+
+
+//TODO: Add links to the Kibana docs for exporting/importing objects.
```
Comment (ruflin) [inline libbeat/docs/breaking.asciidoc]: Want to resolve it as part of this PR?

Comment (dedemorton) [reply libbeat/docs/breaking.asciidoc]: @ruflin If you think it's good enough to merge, maybe we should merge and then iterate over it after it's merged. I think @tsg wanted to review this.

File: libbeat/docs/breaking.asciidoc [hunk 14]
```diff
@@ -9,9 +9,155 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
 
 See the following topics for a description of breaking changes:
 
+* <<breaking-changes-6.3>>
```
Comment (ppf2) [inline libbeat/docs/breaking.asciidoc]: For completeness, I think we will want to include headings for 6.1 and 6.2 as well since there were also breaking changes in these releases that are currently only documented in the release notes today.

https://www.elastic.co/guide/en/beats/libbeat/6.2/release-notes-6.1.0.html#_breaking_changes_3
https://www.elastic.co/guide/en/beats/libbeat/6.2/release-notes-6.2.0.html#_breaking_changes_2
https://www.elastic.co/guide/en/beats/libbeat/6.2/release-notes-6.2.3.html#_breaking_changes

File: libbeat/docs/breaking.asciidoc [hunk 15]
```diff
@@ -9,9 +9,155 @@ changes, but there are breaking changes between major versions (e.g. 5.x to
...
+plugin] documentation. This results in a `host` field in 6.2 indices, and a
+`host.name` field in 6.3 indices. There are no mapping conflicts. Cross index
+searches in Elasticsearch still work as expected. 
+
+*What do you need to change?*
+
+If you searched for the `host` field previously, you now need to search for the
+`host` and `host.name` fields.  
+
+If you have multiple visualizations in Kibana that reference the `host` field,
```
Comment (ppf2) [inline libbeat/docs/breaking.asciidoc]: I will send a separate note on this shortly to discuss.  Somewhere in here, we will want to provide additional details on tackling the outcome of having multiple indices with different schemas for the Kibana use case given Kibana's inconsistent handling of conflicting field types today.  Stay tuned.  Just adding a comment here as something to refine.

File: libbeat/docs/breaking.asciidoc (hunk 14 above)
Comment (ruflin) [reply libbeat/docs/breaking.asciidoc]: @ppf2 I think the difference with the aboves is that these we features which still were in beta.

Comment (dedemorton) [reply libbeat/docs/breaking.asciidoc]: I'll either add sections or make it clear that this section only covers the main breaking changes. We say that in a couple of places plus provide links to the relnotes, but maybe it's not clear enough.

Comment (ppf2) [reply libbeat/docs/breaking.asciidoc]: +1, let's clarify.  Are we saying that the breaking changes guide will only cover breaking changes against the GA product?  Or do we mean that it only covers the "main" breaking changes of GA product?  If "main" is the keyword here, let's also confirm that this is consistent with the rest of the Elastic stack.  Afaik, Elasticsearch documents all breaking changes against the GA product, not just the main ones.

If we are only covering breaking changes of the GA product, we can state that for breaking changes against pre-GA product releases, these are documented only in the release notes, etc.. 

thanks!

Comment (dedemorton) [reply libbeat/docs/breaking.asciidoc]: That's a good question, and I'm not sure about the answer. I'd like @tsg or @ruflin to weigh in here.

**Comment by dedemorton** [comment]
I've created an issue to track the remaining work for this PR (just in case things get busy...I do plan to get this resolved asap): https://github.com/elastic/beats/issues/7408

File: libbeat/docs/breaking.asciidoc (hunk 14 above)
Comment (ruflin) [reply libbeat/docs/breaking.asciidoc]: Moving forward we should document here all breaking changes which happen for any GA features. In general no breaking changes in minor should happen, but there are exceptions.

For beta or experimental features, the breaking changes should only be in the changelog.

**Comment by ppf2** [comment]
@dedemorton @ruflin Can we merge this change to both 6.3 and 6.x branches?  I only see it in master right now :)

**Comment by dedemorton** [comment]
@ppf2 I'll cherry-pick the changes today. Yesterday I was waiting for the doc build on master to be reflected in the published docs, but there was a problem with the site updating.
